File: top-interview-150/207.course-schedule.kt
class Solution {
    fun canFinish(numCourses: Int, prerequisites: Array<IntArray>): Boolean {
        val graph = mutableMapOf<Int, MutableList<Int>>()

        for ((a, b) in prerequisites) {
            graph.getOrPut(a) { mutableListOf<Int>() }.add(b)
        }

        // 0: unvisit, 1: visiting, 2: visited
        val states = mutableMapOf<Int, Int>()
        fun hasCycle(i: Int): Boolean {
            if (states[i] == 1) return true
            if (states[i] == 2) return false

            states[i] = 1
            for (prereq in graph[i] ?: emptyList()) {
                if (hasCycle(prereq)) return true
            }
            states[i] = 2
            return false
        }

        for (i in 0 until numCourses) {
            if (hasCycle(i)) {
                return false
            }
        }
        return true
    }
}

File: top-interview-150/207.course-schedule.py
from typing import List
from collections import defaultdict


class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        graph = defaultdict(list)
        for a, b in prerequisites:
            graph[a].append(b)

        visited = [0] * numCourses

        def hasCycle(i: int) -> bool:
            if visited[i] == 1:
                return True
            if visited[i] == 2:
                return False

            visited[i] = 1
            for j in graph[i]:
                if hasCycle(j):
                    return True
            visited[i] = 2

            return False

        for i in range(numCourses):
            if hasCycle(i):
                return False
        return True
